wait-online: add maximum operational state option

DaanDeMeyer opened the pull request:
Fixes #14529.

yuwata commented on src/libsystemd/sd-network/network-util.c:
```diff
@@ -56,3 +56,54 @@ static const char* const link_address_state_table[_LINK_ADDRESS_STATE_MAX] = {
+
+        if (isempty(min + 1))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                        "Minimum operational state is empty.");
```
Hmm? Sorry, I cannot understand this block.

DaanDeMeyer replied:
Seems I forgot to remove this when I extracted `parse_operational_state` to a separate function.

yuwata commented on src/libsystemd/sd-network/network-util.c:
```diff
@@ -56,3 +56,54 @@ static const char* const link_address_state_table[_LINK_ADDRESS_STATE_MAX] = {
+                                        "Invalid operational state '%s'", str);
+
+        return s;
+}
```
This can be dropped. Please simply use `link_operstate_from_string()`.

yuwata commented on src/libsystemd/sd-network/network-util.c:
```diff
@@ -56,3 +56,54 @@ static const char* const link_address_state_table[_LINK_ADDRESS_STATE_MAX] = {
+
+                state.max = parse_operational_state(p + 1);
+                if (state.max < 0)
+                        return state.max;
```
`return -EINVAL;`

yuwata commented on src/libsystemd/sd-network/network-util.c:
```diff
@@ -56,3 +56,54 @@ static const char* const link_address_state_table[_LINK_ADDRESS_STATE_MAX] = {
+                        return state.max;
+        } else {
+                min = strdup(str);
+                state.max = _LINK_OPERSTATE_INVALID;
```
Maybe, `LINK_OPERSTATE_ROUTABLE`.

DaanDeMeyer replied:
I think `_LINK_OPERSTATE_INVALID` is better in this case. The advantage is that we handle operstates exactly the same in `manager_link_is_online` (use required operstate if it is larger than zero). If the max required operstate is invalid we simply take the link max operstate which defaults to `LINK_OPERSTATE_ROUTABLE`.

DaanDeMeyer commented:
I also renamed `LinkOperationalStatePair` to `LinkOperationalStateRange`.

DaanDeMeyer commented:
In general, I find it hard to figure out when I should log and when I should simply return an error. Are there any guidelines on when errors should be logged and when an error should simply be returned (and leave logging to the caller)?

yuwata commented:
LGTM. Let's wait for the CIs.

> In general, I find it hard to figure out when I should log and when I should simply return an error. Are there any guidelines on when errors should be logged and when an error should simply be returned (and leave logging to the caller)?

In general, the library functions should not log anything except for debugging logs, and caller should logs about the failure.

yuwata commented:
I noticed that this does not update `networkd` part. Maybe, `link_save()` needs to be updated.

DaanDeMeyer commented:
- Changed the operational state range parser to use the default value if the operstate string is empty.
- Introduced `OPERSTATE_RANGE_DEFAULT` to avoid repetition in specifying the default operstate range everywhere.
- Updated the networkd part.

DaanDeMeyer commented:
Reverted the change to the operstate range parser. We don't need the default values in all cases where we use the parser. Instead, we use `_LINK_OPERSTATE_INVALID` which we check at the call site and apply the correct values (depending on the call site).

DaanDeMeyer commented:
One remaining issue is that waiting for interfaces to go down sometimes also involves waiting for the interfaces to disappear which can happen when using veth interfaces with containers and network namespaces. Currently, `systemd-networkd-wait-online -i <interface>:off:off` blocks when the interface is removed because it waits for the interface to appear.

One way to accomodate this would be to add another operational state below `LINK_OPERSTATE_OFF` that signifies a link not being available (`LINK_OPERSTATE_GONE` for example). That could be specified on the command line and we could check for if the link was not found in `manager_configured`. I'm not sure if this is the best solution however, maybe there's a better one.

DaanDeMeyer commented:
I think the remaining CI failure is not caused by this PR.

yuwata commented on src/network/networkd-link.c:
```diff
@@ -3879,8 +3879,9 @@ int link_save(Link *link) {
-                        strempty(link_operstate_to_string(link->network->required_operstate_for_online)));
+                fprintf(f, "REQUIRED_OPER_STATE_FOR_ONLINE=%s:%s\n",
+                        strempty(link_operstate_to_string(link->network->required_operstate_for_online.min)),
+                        strempty(link_operstate_to_string(link->network->required_operstate_for_online.max)));
```
For backward compatibility, please keep the current form when the maximum operstate is not specified. So, maybe, the default value should be `{ LINK_OPERSTATE_DEGRADED, _LINK_OPERSTATE_INVALID }`.

DaanDeMeyer replied:
I kept backwards compatibility by only printing the max value if it does not equal the default value. Just using `_LINK_OPERSTATE_INVALID` and `strempty` is not sufficient since we'd still be printing the `:` character so we have to do the `if` regardless.

DaanDeMeyer commented:
CI failures do not seem related.

mrc0mmand commented:
> CI failures do not seem related.

Yup, the failing `test-execute` is caused by #14548. I'll reschedule the failing jobs once I deploy a temporary quick-fix.

DaanDeMeyer commented:
I added an extra commit to allow waiting for interfaces to disappear based on the approach I mentioned previously (I named the new operational state `LINK_OPERSTATE_MISSING`). Let me know if this approach is fine or if we should move the support for waiting for interfaces to disappear to a separate PR.

keszybz commented on src/libsystemd/sd-network/network-util.c:
```diff
@@ -56,3 +56,48 @@ static const char* const link_address_state_table[_LINK_ADDRESS_STATE_MAX] = {
+
+        if (range.min != _LINK_OPERSTATE_INVALID &&
+                range.max != _LINK_OPERSTATE_INVALID &&
+                range.min > range.max)
```
Indentatino.

DaanDeMeyer replied:
I'm assuming you meant continuation indents should also be 8 spaces. Please let me know if you meant something else (I looked at the style guide but didn't immediately see anything about continuation indents).

keszybz replied:
We just generally follow what emacs does.  In this case, lines 96–97 should be indented to ling up with "range" above.

keszybz commented on src/network/networkd-network.c:
```diff
@@ -1321,11 +1321,15 @@ int config_parse_required_for_online(
+        if (range.min != _LINK_OPERSTATE_INVALID)
+                network->required_operstate_for_online.min = range.min;
+        if (range.max != _LINK_OPERSTATE_INVALID)
+                network->required_operstate_for_online.max = range.max;
```
This seems wrong. If I do:
```
RequiredForOnline=A:B
RequiredForOnline=:C
```
then the first assignment should be completely overwritten by the second one. With the two `if`s above, I'll get something like A:C at the end.

DaanDeMeyer replied:
Changed the logic to explicitly set the default value again if an operstate is empty.

keszybz commented on src/network/wait-online/link.c:
```diff
@@ -122,18 +121,26 @@ int link_update_monitor(Link *l) {
+                        if (range.min != _LINK_OPERSTATE_INVALID)
+                                l->required_operstate.min = range.min;
+                        if (range.max != _LINK_OPERSTATE_INVALID)
+                                l->required_operstate.max = range.max;
```
Here too.

poettering commented:
Hmm, this cements that there was an ordering between the operational states. is there really? isn't parts of it not more multi-dimensional?

Shouldn't this be a configurable set of allowed states? I mean, the min/max stuff is mostly just a shortcut for a set here, no? maybe be safe here and really make it a set and expect people to list all states?

DaanDeMeyer commented:
I kinda agree but I'm not sure how to handle backwards compatibility if we do that with the same option (`--interface`) since it currently accepts the minimum acceptable state. To keep backwards compatibility, specifying only a single state would result in the set consisting of that state and all 'higher' ones. The question then becomes: How do we specify a single state to wait for?

I'm all for breaking backwards compatibility since I'd only be relying on the new behaviour but I imagine there will be some scripts out there depending on the current behaviour.

keszybz commented:
The set of states is *mostly* linear. There are few cases which can't be strictly ordered. But I think it is a reasonable compromise to linear ordering, because it is much easier for the user to specify a min or max state rather than having to list all intermediate states explicitly. I expect users will mostly use this as :carrier, or as degraded:, without caring at all that the ordering between "dormant" and "degraded-carrier" is somewhat arbitrary. Min:max also covers the original request. So I think current approach is OK. We can always extend this in the future if desired.

(One possible way to extend in a backwards compatible way: if only one item is present, treat as now. If at least one coma is present, treat single items as single items, and any ranges as currently.)

keszybz commented on src/network/networkd-network.c:
```diff
@@ -1321,11 +1321,17 @@ int config_parse_required_for_online(
+                                                                : LINK_OPERSTATE_RANGE_DEFAULT.min;
+        network->required_operstate_for_online.max = range.max != _LINK_OPERSTATE_INVALID
+                                                                ? range.max
+                                                                : LINK_OPERSTATE_RANGE_DEFAULT.max;
```
This is still wrong. ":dormant" should mean anything between "missing" (the lowest state) and "dormant", but right now this is the same as "degraded:dormant" (which doesn't even make sense). I.e. please just do the same change as in the other place.

DaanDeMeyer replied:
I moved the default value logic into `parse_operational_state_range`. The only thing it doesn't handle is empty operational state ranges as those have to be handled differently depending on whether we're parsing unit code or operational state ranges passed on the command line. In unit code, we default to degraded:routable if an operational state range is empty. In wait-online, we error on empty operational state ranges. (`-i eth1` is fine, `-i eth1:` is not)

keszybz commented on src/network/wait-online/link.c:
```diff
@@ -122,18 +121,28 @@ int link_update_monitor(Link *l) {
+                                                                : LINK_OPERSTATE_RANGE_DEFAULT.min;
+                        l->required_operstate.max = range.max != _LINK_OPERSTATE_INVALID
+                                                                ? range.max
+                                                                : LINK_OPERSTATE_RANGE_DEFAULT.max;
```
Same here.

keszybz commented on src/network/wait-online/wait-online.c:
```diff
@@ -152,14 +152,17 @@ static int parse_argv(int argc, char *argv[]) {
+                        if (r < 0)
+                                return log_error_errno(r, "Invalid operational state range '%s'", optarg);
+
+                        if (range.min == _LINK_OPERSTATE_INVALID && range.max == _LINK_OPERSTATE_INVALID)
```
This check can go away. parse_operational_state_range() should always return a valid range.

DaanDeMeyer replied:
Changed `parse_operational_state_range` to fail on empty strings. Checks on empty strings now happen before calling `parse_operational_state_range` where necessary.

DaanDeMeyer commented:
@keszybz PTAL, I think everything is in order now.

DaanDeMeyer commented:
Ping, would be great if this could be merged or to know if further modifications are needed. I have some scripts that build systemd with this PR that I would like to point to the official repository instead of my fork.

keszybz commented:
I think this is all good now, let's merge.
